core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -322,8 +332,9 @@ public Node start() {
 
         validateNodeBeforeAcceptingRequests(settings, transportService.boundAddress());
 
-        DiscoveryNode localNode = injector.getInstance(DiscoveryNodeService.class)
-                .buildLocalNode(transportService.boundAddress().publishAddress());
+        String nodeId = injector.getInstance(NodeEnvironment.class).nodeId();
```

Comment:
`NodeEnvironment` is created in the constructor and `localNode` injection happens in the `start` method. I could make `nodeEnvironment` or `nodeId` a final field in the `Node` class, but I'm not sure that makes the code much nicer.